plugins/modem-manager/fu-mbim-qdu-updater.c

```diff
@@ -0,0 +1,545 @@
+						0x00,	/* start addr */
+						0x00,	/* page_sz */
+						out_max_transfer_size);
+	chk = g_ptr_array_index (ctx->chunks, 0);
+	request = (mbim_message_qdu_file_write_set_new (fu_chunk_get_data_sz (chk), (const guint8 *)fu_chunk_get_data (chk), &error));
+	mbim_device_command (ctx->mbim_device,
+			     request,
+			     10,
+			     ctx->cancellable,
+			     (GAsyncReadyCallback)fu_mbim_qdu_updater_file_write_ready,
+			     ctx);
+	mbim_message_unref (request);
+}
+
+static void
+fu_mbim_qdu_updater_session_ready (MbimDevice *device, GAsyncResult *res, gpointer user_data)
+{
+	MbimMessage *response = NULL;
+	MbimQduSessionType     out_current_session_type;
+	MbimQduSessionStatus   out_current_session_status;
+	MbimQduSessionType     out_last_session_type;
+	MbimQduSessionResult   out_last_session_result;
+	guint32                out_last_session_error_offset;
+	guint32                out_last_session_error_size;
+	g_autoptr(GError)      error = NULL;
+	MbimMessage *request = NULL;
+	WriteContext *ctx = user_data;
+
+	response = mbim_device_command_finish (device, res, &error);
+	if (!response || !mbim_message_response_get_result (response, MBIM_MESSAGE_TYPE_COMMAND_DONE, &error)) {
+		g_printerr ("error: operation failed: %s\n", error->message);
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	if (!mbim_message_qdu_update_session_response_parse (
+							     response,
+							     &out_current_session_type,
+							     &out_current_session_status,
+							     &out_last_session_type,
+							     &out_last_session_result,
+							     &out_last_session_error_offset,
+							     &out_last_session_error_size,
+							     &error)) {
+		g_printerr ("error: couldn't parse response message: %s\n", error->message);
+		g_printerr ("out_current_session_type: %d\n", (int)out_current_session_type);
+		g_printerr ("out_current_session_status: %d\n", (int)out_current_session_status);
+		g_printerr ("out_last_session_type: %d\n", (int)out_last_session_type);
+		g_printerr ("out_last_session_result: %d\n", (int)out_last_session_result);
+		g_printerr ("out_last_session_error_offset: %d\n", (int)out_last_session_error_offset);
+		g_printerr ("out_last_session_error_size: %d\n", (int)out_last_session_error_size);
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	if (response)
+		mbim_message_unref (response);
+
+	g_debug ("[%s] Successfully request modem to update session", mbim_device_get_path_display (device));
+
+	request = (mbim_message_qdu_file_open_set_new (0, g_bytes_get_size (ctx->blob), &error));
+	mbim_device_command (device,
+			     request,
+			     10,
+			     ctx->cancellable,
+			     (GAsyncReadyCallback)fu_mbim_qdu_updater_file_open_ready,
+			     ctx);
+	mbim_message_unref (request);
+}
+
+static void
+fu_mbim_qdu_updater_set_update_session (MbimDevice *device, WriteContext *ctx)
+{
+  	MbimMessage *request = NULL;
+
+	request = (mbim_message_qdu_update_session_set_new (0, 1, &ctx->error));
```

Comment:
You can assume the set_new() always succeeds. Just pass a NULL for the GError and don't check for errors.